### PR TITLE
[Datastore] Refactor internal get_filesystem() method

### PR DESCRIPTION
**Pull request by alxtkr77**
https://jira.iguazeng.com/browse/ML-5510
Refactoring an internal get_filesystem() method

### CODE REVIEW

File: mlrun/datastore/v3io.py [hunk 1]
```diff
@@ -73,18 +73,16 @@ def url(self):
...
             import v3iofs  # noqa
         except ImportError as exc:
-            if not silent:
-                raise ImportError(
-                    "v3iofs or storey not installed, run pip install storey"
-                ) from exc
-            return None
+            raise ImportError(
+                "v3iofs or storey not installed, run pip install storey"
+            ) from exc
```
Comment (gtopper) [inline mlrun/datastore/v3io.py:85]: `v3iofs` is a base requirement of mlrun. What does this have to do with storey?

File: mlrun/datastore/s3.py [hunk 2]
```diff
@@ -109,18 +109,14 @@ def get_spark_options(self):
...
             return self._filesystem
         try:
             import s3fs  # noqa
         except ImportError as exc:
-            if not silent:
-                raise ImportError(
-                    "AWS s3fs not installed, run pip install s3fs"
-                ) from exc
-            return None
+            raise ImportError("AWS s3fs not installed, run pip install s3fs") from exc
```
Comment (gtopper) [inline mlrun/datastore/s3.py:119]:
```suggestion
            raise ImportError("AWS s3fs not installed") from exc
```
The user should _not_ install the latest s3fs. They should install `mlrun[complete]` or  `mlrun[s3]` with a version that is compatible with the mlrun service on their env.

File: mlrun/datastore/azure_blob.py [hunk 3]
```diff
@@ -86,12 +83,12 @@ def _convert_key_to_remote_path(self, key):
 
     def upload(self, key, src_path):
         remote_path = self._convert_key_to_remote_path(key)
-        self._filesystem.put_file(src_path, remote_path, overwrite=True)
+        self.get_filesystem().put_file(src_path, remote_path, overwrite=True)
```
Comment (gtopper) [inline mlrun/datastore/azure_blob.py:86]: I think if we're going to replace it everywhere, it makes sense to make it a `@property`.

File: mlrun/datastore/azure_blob.py [hunk 4]
```diff
@@ -33,20 +33,17 @@ class AzureBlobStore(DataStore):
...
         try:
             import adlfs  # noqa
         except ImportError as exc:
-            if not silent:
-                raise ImportError(
-                    f"Azure adlfs not installed, run pip install adlfs, {err_to_str(exc)}"
-                )
-            return None
+            raise ImportError(
+                f"Azure adlfs not installed, run pip install adlfs, {err_to_str(exc)}"
```
[the remark this answers is not in the document]
Comment (gtopper) [reply mlrun/datastore/azure_blob.py:45]: I realize it wasn't introduced in this PR, though.

File: mlrun/datastore/s3.py (hunk 2 above)
Comment (gtopper) [reply mlrun/datastore/s3.py:119]: I realize it wasn't introduced in this PR, though.

File: mlrun/datastore/v3io.py (hunk 1 above)
Comment (gtopper) [reply mlrun/datastore/v3io.py:85]: I realize it wasn't introduced in this PR, though.

File: mlrun/datastore/azure_blob.py (hunk 3 above)
Comment (alxtkr77) [reply mlrun/datastore/azure_blob.py:86]: get_filesystem() is a public method in the DataStore class and is overridden in subclasses as part of the public interface. Would it be appropriate to convert it into a @property instead?

**Review by gtopper — APPROVED** [review]
Looks awesome now!

File: mlrun/datastore/azure_blob.py (hunk 3 above)
Comment (gtopper) [reply mlrun/datastore/azure_blob.py]: I don't think it's really meant to be user facing, but if in doubt, consider adding this to the base class:
```python
def get_filesystem(self):
    return self.filesystem
```

Comment (assaf758) [reply mlrun/datastore/azure_blob.py]: Let's add it with deprecation warning

File: mlrun/datastore/base.py [hunk 5]
```diff
@@ -91,6 +92,16 @@ def uri_to_kfp(endpoint, subpath):
     def uri_to_ipython(endpoint, subpath):
         return ""
 
+    # TODO: remove in 1.8.0
+    @deprecated(
+        version="1.8.0",
+        reason="'get_filesystem()' will be removed from this file in 1.8.0, use "
```
Comment (assaf758) [inline mlrun/datastore/base.py:98]: Please remove "from this file"
